inventory.lua

```diff
@@ -83,63 +106,203 @@ minetest.register_node("digilines:chest", {
+			-- Here, direction = direction item is moving, which is into side.
+			local side = vector.multiply(direction, -1)
 			local inv = minetest.get_meta(pos):get_inventory()
-			local leftover = inv:add_item("main", stack)
-			local count = leftover:get_count()
-			if count == 0 then
-				local derpstack = stack:get_name()..' 1'
-				if not inv:room_for_item("main", derpstack) then
-					-- when you can't put a single more of whatever you just put,
-					-- you'll get a put for it, then a full
-					sendMessage(pos,"full "..maybeString(stack)..' '..tostring(count))
+			local inv_contents = inv:get_list("main")
+			local any_put = false
+			local stack = original_stack
+			local stack_name = stack:get_name()
+			local stack_count = stack:get_count()
+			-- Walk the inventory, adding items to existing stacks of the same
+			-- type.
+			for i = 1, #inv_contents do
+				local existing_stack = inv_contents[i]
+				if not existing_stack:is_empty() and existing_stack:get_name() == stack_name then
+					local leftover = existing_stack:add_item(stack)
+					local leftover_count = leftover:get_count()
+					if leftover_count ~= stack_count then
+						-- We put some items into the slot. Update the slot in
+						-- the inventory, tell Digilines listeners about it,
+						-- and keep looking for the a place to put the
+						-- leftovers if any.
+						any_put = true
+						inv:set_stack("main", i, existing_stack)
+						local stack_that_was_put
+						if leftover_count == 0 then
+							stack_that_was_put = stack
+						else
+							stack_that_was_put = ItemStack(stack)
+							stack_that_was_put:set_count(stack_count - leftover_count)
+						end
+						send_message(pos, "tput", stack_that_was_put, nil, i, side)
+						stack = leftover
+						stack_count = leftover_count
+						if stack_count == 0 then
+							break
+						end
+					end
+				end
+			end
+			if stack_count ~= 0 then
+				-- Walk the inventory, adding items to empty slots.
+				for i = 1, #inv_contents do
+					local existing_stack = inv_contents[i]
+					if existing_stack:is_empty() then
+						local leftover = existing_stack:add_item(stack)
+						local leftover_count = leftover:get_count()
+						if leftover_count ~= stack_count then
+							-- We put some items into the slot. Update the slot in
+							-- the inventory, tell Digilines listeners about it,
+							-- and keep looking for the a place to put the
+							-- leftovers if any.
+							any_put = true
+							inv:set_stack("main", i, existing_stack)
+							local stack_that_was_put
+							if leftover_count == 0 then
+								stack_that_was_put = stack
+							else
+								stack_that_was_put = ItemStack(stack)
+								stack_that_was_put:set_count(stack_count - leftover_count)
+							end
+							send_message(pos, "tput", stack_that_was_put, nil, i, side)
+							stack = leftover
+							stack_count = leftover_count
+							if stack_count == 0 then
+								break
+							end
+						end
+					end
```

Comment:
I’m not  a huge fan of this code. It replicates the logic of `InventoryRef:add_item` (which doesn’t tell you which slot(s) it decided to use), but it’s also a lot of code. I’m considering changing it to call `InventoryRef:add_item` like it did before, but use `InventoryRef:get_list` to take a snapshot before and after the operation and then diff the stacks to find out where the items were put. I think the code would be a lot smaller. It might also be somewhat faster, due to offloading more of the logic to C++, only iterating the slots once in Lua, and probably doing fewer calls to native functions, but I’m not sure about that; it could also be slower because of constructing the (large) stack lists twice. Any thoughts?